pal/src/httpapi_sl.c

```diff
@@ -0,0 +1,492 @@
+            if ((stringcasecmp(hname, "content-length") != 0) &&
+                    (stringcasecmp(hname, "host") != 0)) {
+                ret = HTTPClient_setHeaderByName(cli,
+                        HTTPClient_REQUEST_HEADER_MASK,
+                        hname, hvalue,
+                        strlen(hvalue) + 1, HTTPClient_HFIELD_NOT_PERSISTENT);
+                if (ret < 0) {
+                    LogError("Failed setting request header, ret=%d", ret);
+                }
+            }
+        }
+
+        free(hname);
+        hname = NULL;
+
+        if (ret < 0) {
+            return (HTTPAPI_SEND_REQUEST_FAILED);
+        }
+    }
+
+    /* Send the request */
+    ret = HTTPClient_sendRequest(cli, method,
+            relativePath, (const char *)content, contentLength, 0);
+    if (ret < 0) {
+        LogError("HTTPClient_sendRequest failed, ret=%d", ret);
+        return (HTTPAPI_SEND_REQUEST_FAILED);
+    }
+
+    *statusCode = (unsigned int)ret;
+
+    /* Get the response headers */
+    cnt = 0;
+    offset = 0;
+
+    /*
+     * TODO: If there is more than one header of the same name, is this
+     * an issue with HTTPClient?
+     */
+    for (i = 0; i <= HTTPClient_MAX_RESPONSE_HEADER_FILEDS; i++) {
+        contentBufLen = CONTENT_BUF_LEN;
+        ret = HTTPClient_getHeader(cli, i, contentBuf, &contentBufLen, 0);
+        if (ret == HTTPClient_EGETOPTBUFSMALL) {
+            /* TODO: content buffer is too small. Enlarge and try again?? */
+            LogError("Content buffer is too small for incoming header");
+            ret = HTTPAPI_HTTP_HEADERS_FAILED;
+            goto headersDone;
+        }
+        else if (ret < 0) {
+            LogError("Failed to get header, ret=%d", ret);
+            ret = HTTPAPI_HTTP_HEADERS_FAILED;
+            goto headersDone;
+        }
+        else if (contentBufLen == 0) {
+            /* No data for this header */
+            continue;
+        }
+
+        ret = HTTPHeaders_AddHeaderNameValuePair(responseHeadersHandle,
+                HEADER_TO_STR(i), contentBuf);
+        if (ret != HTTP_HEADERS_OK) {
+            LogError("Adding the response header failed");
+            ret = HTTPAPI_HTTP_HEADERS_FAILED;
+            goto headersDone;
+        }
+        offset = 0;
+    }
+
+headersDone:
+    hname = NULL;
+    if (ret != 0) {
+        return ((HTTPAPI_RESULT)ret);
+    }
+
+    /* Get response body */
+    if (responseContent != NULL) {
+        offset = 0;
+        cnt = 0;
+
+        do {
+            ret = HTTPClient_readResponseBody(cli, contentBuf, CONTENT_BUF_LEN,
+                    &moreFlag);
+
+            if (ret < 0) {
+                LogError("HTTP read response body failed, ret=%d", ret);
+                ret = HTTPAPI_RECEIVE_RESPONSE_FAILED;
+                goto contentDone;
+            }
+
+            if (ret != 0) {
+                cnt = ret;
+                ret = BUFFER_enlarge(responseContent, cnt);
+                if (ret != 0) {
```

Comment:
It would be more efficient to check the `Content-Length` and `Transfer-Encoding` headers and size the buffer once in advance when possible